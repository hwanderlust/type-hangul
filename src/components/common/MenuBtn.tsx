import React from "react";
import { Link } from "react-router-dom";
import styled from "styled-components";

import { Fonts, StyledProps } from "../../helpers";
import icon from "../../images/apeach-btn.png";

interface WindowProps {
  numOfItems: number;
}
const Window = styled.div<WindowProps>`
  width: 100px;
  display: flex;
  justify-content: center;
  position: absolute;
  height: ${props => props.numOfItems * 30}px;
`;
const TopWindow = styled(Window)`
  bottom: 100%;
`;
const List = styled.ul`
  list-style-type: none;
  margin: 0;
  padding: 0;
  display: flex;
  flex-direction: column;
  justify-content: space-around;
  align-items: center;
`;
const Button = styled.button`
  width: 100px;
  height: 100px;
  background: transparent;
  border: none;
  border-radius: 50%;
  cursor: pointer;
  filter: drop-shadow(1px 0 10px #000);
  
  &:focus {
    border: 1px solid white;
    outline: none;
  }

  @media only screen and (max-width: 720px) {
    width: 50px;
    height: 50px;
  }
`;
const Icon = styled.img`
  width: 100%;
  height: 100%;
`;

const topOptions = ["run", "pop", "jump"];
const bottomOptions = ["about", "contact"];
const MenuOption = styled(Link)`
  font-family: ${Fonts.playfair};
  font-size: 12px;
  text-decoration: none;
  text-transform: uppercase;
  color: white;
  filter: drop-shadow(1px 0 5px #000);

  &:hover {
    color: black;
  }
`;

interface State {
  isOpen: boolean;
}

class MenuBtn extends React.Component<StyledProps, State> {
  private btn: React.RefObject<HTMLButtonElement>;

  constructor(props: StyledProps) {
    super(props);
    this.state = { isOpen: false };
    this.btn = React.createRef();
  }

  handleClick = (e: React.MouseEvent<HTMLButtonElement, MouseEvent>) => {
    e.preventDefault();
    this.setState((prevState: State) => ({ isOpen: !prevState.isOpen }));
    (this.btn.current as HTMLButtonElement).blur();
  }

  render() {
    return (
      <div {...this.props}>
        {this.state.isOpen && (
          <TopWindow numOfItems={topOptions.length}>
            <List>
              {topOptions.map(option => <MenuOption to={`/game/${option}`} key={`top-${option}`}>{option}</MenuOption>)}
            </List>
          </TopWindow>
        )}

        <Button ref={this.btn} onClick={this.handleClick}>
          <Icon src={icon} alt="Kakao Friends' Apeach smiling - as a menu icon/button" />
        </Button>

        {this.state.isOpen && (
          <Window numOfItems={bottomOptions.length}>
            <List>
              {bottomOptions.map(option => <MenuOption to={`/${option}`} key={`bottom-${option}`}>{option}</MenuOption>)}
            </List>
          </Window>
        )}
      </div>
    );
  }
}

export default MenuBtn;