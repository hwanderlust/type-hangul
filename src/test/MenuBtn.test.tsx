import 'jest-styled-components';
import React from "react";
import ReactDOM from "react-dom";
import ReactTestUtils from "react-dom/test-utils";
import { BrowserRouter } from "react-router-dom";
import renderer from "react-test-renderer";

import MenuBtn from "../components/common/MenuBtn";

describe("<MenuBtn/>", () => {
  it("renders correctly in its closed state", () => {
    const tree = renderer
      .create(
        <BrowserRouter>
          <MenuBtn />
        </BrowserRouter>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("renders correctly in its opened state", () => {
    const container = document.createElement("div");
    ReactTestUtils.act(() => {
      ReactDOM.render(
        <BrowserRouter>
          <MenuBtn />
        </BrowserRouter>, container);
    });

    const button = container.querySelector("button");
    if (button) {
      ReactTestUtils.Simulate.click(button);
      expect(container.innerHTML).toMatchSnapshot();
    }
  });
});
