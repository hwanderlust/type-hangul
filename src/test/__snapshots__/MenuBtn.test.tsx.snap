// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MenuBtn/> renders correctly in its closed state 1`] = `
.c0 {
  width: 100px;
  height: 100px;
  background: transparent;
  border: none;
  border-radius: 50%;
  cursor: pointer;
  -webkit-filter: drop-shadow(1px 0 10px #000);
  filter: drop-shadow(1px 0 10px #000);
}

.c0:focus {
  border: 1px solid white;
  outline: none;
}

.c1 {
  width: 100%;
  height: 100%;
}

@media only screen and (max-width:720px) {
  .c0 {
    width: 50px;
    height: 50px;
  }
}

<div>
  <button
    className="c0"
    onClick={[Function]}
  >
    <img
      alt="Kakao Friends' Apeach smiling - as a menu icon/button"
      className="c1"
      src="apeach-btn.png"
    />
  </button>
</div>
`;

exports[`<MenuBtn/> renders correctly in its opened state 1`] = `"<div><div class=\\"sc-bdnylx sc-gtssRu egspcC hUggKH\\"><ul class=\\"sc-dlnjPT foHzqh\\"><a class=\\"sc-jSFkmK bYQpRM\\" href=\\"/game/pop\\">pop</a><a class=\\"sc-jSFkmK bYQpRM\\" href=\\"/game/jump\\">jump</a></ul></div><button class=\\"sc-hKFyIo cvRZQX\\"><img src=\\"apeach-btn.png\\" alt=\\"Kakao Friends' Apeach smiling - as a menu icon/button\\" class=\\"sc-eCApGN fGjSpM\\"></button><div class=\\"sc-bdnylx egspcC\\"><ul class=\\"sc-dlnjPT foHzqh\\"><a class=\\"sc-jSFkmK bYQpRM\\" href=\\"/about\\">about</a><a class=\\"sc-jSFkmK bYQpRM\\" href=\\"/contact\\">contact</a></ul></div></div>"`;
